algebra-core/src/curves/cuda/scalar_mul/cpu_gpu_macros.rs

```diff
@@ -0,0 +1,252 @@
+// TODO: make this more generic
+#[macro_export]
+macro_rules! impl_gpu_cpu_run_kernel {
+    () =>  {
+        fn clear_gpu_profiling_data() {
+            #[cfg(feature = "cuda")]
+            {
+                let dir = dirs::cache_dir()
+                    .unwrap()
+                    .join("zexe-algebra")
+                    .join("cuda-scalar-mul-profiler")
+                    .join(P::namespace());
+                std::fs::create_dir_all(&dir).expect("Could not create/get cache dir for profile data");
+                std::fs::File::create(&dir.join("profile_data.txt")).expect("could not create profile_data.txt");
+            }
+        }
+
+        /// We split up the job statically between the CPU and GPUs
+        /// based on continuous profiling stored both in a static location in memory
+        /// that is lost the moment the progam stops running.
+        /// and also a txt file in the OS' cache dir.
+
+        /// Only one such procedure should be running at any time.
+        #[allow(unused_variables)]
+        fn cpu_gpu_static_partition_run_kernel(
+            bases_h: &mut [<Self as ProjectiveCurve>::Affine],
+            exps_h: &[<<Self as ProjectiveCurve>::ScalarField as PrimeField>::BigInt],
+            cuda_group_size: usize,
+            // size of the batch for cpu scalar mul
+            cpu_chunk_size: usize,
+        ) {
+            #[cfg(feature = "cuda")]
+            {
+                if !Device::init() {
+                    panic!("Do not call this function unless the device has been checked to initialise successfully");
+                }
+                let n_devices = Device::get_count().unwrap();
+                let n = bases_h.len();
+                // Create references so we can split the slices
+                let mut res_ref = &mut bases_h[..];
+                let mut exps_h_ref = exps_h;
+
+                let _now = timer!();
+                // Get data for proportion of total throughput achieved by each device
+                let dir = dirs::cache_dir()
```

Comment:
Is profiling going to be always run even in normal usage?

Reply:
I'm not sure what you mean by normal usage. I have yet to expose strictly GPU-only functionality. The plan is to hide internal functions using a `GPUScalarMulInternal` trait defined only at the `pub(crate)` level to hide these functions.

Reply:
Got it - so who's using these? Only people who want to run one-off profiling?

Reply:
Further, the profiling state is saved between runs. So its essentially a persistent log.

Reply:
What I was asking initially is "normal usage", meaning "is this function called when running, let's say, snark setup". So when is that function called?

Reply:
Does the GPU code use it to tune some parameters for the future? What's the goal of the function?